src/ddsrt/src/sync/freertos/sync.c

```diff
@@ -0,0 +1,505 @@
+}
+
+static bool
+mutex_lock(ddsrt_mutex_t *mutex, int blk)
+{
+  assert(mutex != NULL);
+
+  if (xSemaphoreTake(mutex->sem, (blk == 1 ? portMAX_DELAY : 0)) != pdPASS) {
+    DDS_TRACE("Failed to lock 0x%p", mutex);
+    /* xSemaphoreTake will only return pdFAIL on timeout. The wait will be
+       indefinite if INCLUDE_vTaskSuspend is set to 1 in FreeRTOSConfig.h
+       and portMAX_DELAY was passed. */
+    assert(blk == 0);
+    return false;
+  }
+
+  return true;
+}
+
+void ddsrt_mutex_lock(ddsrt_mutex_t *mutex)
+{
+  if (!mutex_lock(mutex, 1)) {
+    abort();
+  }
+}
+
+bool ddsrt_mutex_trylock(ddsrt_mutex_t *mutex)
+{
+  return mutex_lock(mutex, 0);
+}
+
+void ddsrt_mutex_unlock(ddsrt_mutex_t *mutex)
+{
+  assert(mutex != NULL);
+
+  if (xSemaphoreGive(mutex->sem) != pdPASS) {
+    DDS_TRACE("Failed to unlock 0x%p", mutex->sem);
+    abort();
+  }
+}
+
+static dds_retcode_t
+cond_resize(ddsrt_cond_t *cond, size_t len)
+{
+  dds_retcode_t rc = 0;
+  size_t off;
+  TaskHandle_t *buf = NULL;
+  static const size_t sz = sizeof(*cond->tasks.buf);
+
+  assert(cond != NULL);
+
+  /* Free buffer if required number of slots is zero. */
+  if (len == 0) {
+    ddsrt_free(cond->tasks.buf);
+    (void)memset(&cond->tasks, 0, sizeof(cond->tasks));
+  } else {
+    if (len > cond->tasks.len) {
+      off = cond->tasks.len;
+    } else {
+      off = len;
+    }
+
+    /* Resize buffer if required number of slots is more than available or
+       amount of space freed up is worth the effort. */
+    if (len > cond->tasks.len || len < (cond->tasks.len - 1)) {
+      buf = ddsrt_realloc(cond->tasks.buf, len * sz);
+      if (buf == NULL) {
+        rc = DDS_RETCODE_OUT_OF_RESOURCES;
+      } else {
+        cond->tasks.buf = buf;
+        cond->tasks.len = len;
+      }
+    }
+
+    /* Initialize memory that should not be used anymore/yet to zero. */
+    if (off < cond->tasks.len) {
+      (void)memset(cond->tasks.buf + off, 0, (cond->tasks.len - off) * sz);
+    }
+  }
+
+  return rc;
+}
+
+static dds_retcode_t
+cond_push(ddsrt_cond_t *cond, TaskHandle_t task)
+{
+  size_t cnt;
+
+  assert(cond != NULL);
+  assert(task != NULL);
+
+  /* Ensure the task is not listed. */
+  for (cnt = 0; cnt < cond->tasks.cnt && cond->tasks.buf[cnt] != task; cnt++) {
+    /* Do nothing. */
+  }
+
+  assert(cnt <= cond->tasks.cnt);
+
+  if (cnt == cond->tasks.len && cond_resize(cond, cnt + 1) != DDS_RETCODE_OK) {
+    return DDS_RETCODE_OUT_OF_RESOURCES;
+  }
+
+  cond->tasks.buf[cond->tasks.cnt] = task;
+  cond->tasks.cnt++;
+
+  return DDS_RETCODE_OK;
+}
+
+static void
+cond_pop(ddsrt_cond_t *cond, TaskHandle_t task)
+{
+  size_t cnt;
+
+  assert(cond != NULL);
+  assert(task != NULL);
+
+  /* Locate task index. */
+  for (cnt = 0; cnt < cond->tasks.cnt && cond->tasks.buf[cnt] != task; cnt++) {
+    /* Do nothing. */
+  }
+
+  if (cnt < cond->tasks.cnt) {
+    cond->tasks.cnt--;
+
+    if (cnt < cond->tasks.cnt) {
+      /* Ensure list of tasks is neatly packed. */
```

Comment:
If ordering doesn't matter (it doesn't seem to), it'd make more sense to swap the last one in. Actually yes, it does matter for fairness.

Still, all those ``memmove``s and reallocating every time ... turning it into a circular queue and shrinking it only when there are many available slots instead of just 1 might be a good idea. This stuff gets called a lot ...

Reply:
Indeed, the ordering is to ensure every task gets it's turn.

I've made the code so that more than one slot must be free for a resize shrink operation, but choosing a larger number might be better. Also, I like the idea about using a circular queue. I'll update the implementation and the pull request.